cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
+                            allocator_type(), stream);
+}
+
+/**
+ * @brief Computes all aggregations from @p requests that require a single pass
+ * over the data and stores the results in @p sparse_results
+ */
+template <bool keys_have_nulls, typename Map>
+void compute_single_pass_aggs(table_view const& keys,
+                              std::vector<aggregation_request> const& requests,
+                              experimental::detail::result_cache& sparse_results,
+                              Map& map, bool ignore_null_keys,
+                              cudaStream_t stream)
+{
+  // flatten the aggs to a table that can be operated on by aggregate_row
+  table_view flattened_values;
+  std::vector<aggregation::Kind> aggs;
+  std::vector<size_t> col_ids;
+  std::tie(flattened_values, aggs, col_ids) = flatten_single_pass_aggs(requests);
+
+  // make table that will hold sparse results
+  std::vector<std::unique_ptr<column>> sparse_columns;
+  for (size_t i = 0; i < aggs.size(); i++) {
+    auto const& col = flattened_values.column(i);
+    bool nullable = (aggs[i] == aggregation::COUNT) ? false : col.has_nulls();
+    auto mask_state = (nullable) ? ALL_NULL : UNALLOCATED;
+    
+    sparse_columns.emplace_back(make_fixed_width_column(
+      experimental::detail::target_type(col.type(), aggs[i]),
+      col.size(), mask_state, stream));
+  }
+  table sparse_table(std::move(sparse_columns));
+  mutable_table_view table_view = sparse_table.mutable_view();
+  experimental::detail::initialize_with_identity(table_view, aggs, stream);
+
+  // prepare to launch kernel to do the actual aggregation
+  auto d_sparse_table = mutable_table_device_view::create(sparse_table);
+  auto d_values = table_device_view::create(flattened_values);
+  rmm::device_vector<aggregation::Kind> d_aggs(aggs);
+
+  bool skip_key_rows_with_nulls = keys_have_nulls and ignore_null_keys;
+
+  experimental::detail::grid_1d grid(keys.num_rows(), 256);
+  if (skip_key_rows_with_nulls) {
+    auto row_bitmask{bitmask_and(keys, rmm::mr::get_default_resource(), stream)};
+    hash::compute_single_pass_aggs<true>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        static_cast<bitmask_type*>(row_bitmask.data()));
+  } else {
+    hash::compute_single_pass_aggs<false>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        nullptr);
+  }
+
+  // Add results back to sparse_results cache
+  auto sparse_result_cols = sparse_table.release();
+  for (size_t i = 0; i < aggs.size(); i++) {
+    sparse_results.add_result(col_ids[i],
+                              std::make_unique<aggregation>(aggs[i]),
+                              std::move(sparse_result_cols[i]));
+  }
+}
+
+template <bool keys_have_nulls>
+auto groupby_null_templated(
```

Comment:
Is auto return type helpful here? It means I can't see the return type without going to the call site. Since the return type is always `unique_ptr<table>`, I think it would be clearer to make it explicit here.